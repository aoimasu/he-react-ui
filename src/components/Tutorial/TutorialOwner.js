// @flow

import * as React from 'react';
import TutorialContext from './TutorialContext';

export default class TutorialOwner extends React.Component<*, *> {
  state = {
    tutorialIndex: this.props.autoStart ? 0 : -1,
  };

  onTutorialAdvance = () => {
    const { steps } = this.props;
    const { tutorialIndex } = this.state;

    const nexttutorialIndex = tutorialIndex + 1;

    this.setState({
      tutorialIndex: nexttutorialIndex >= steps.length ? -1 : nexttutorialIndex,
    });
  };

  onTutorialDismiss = () => {
    this.setState({ tutorialIndex: -1 });
  };

  onTutorialJump = (id: string) => {
    const tutorialIndex = this.props.steps.findIndex(it => it === id);
    this.setState({ tutorialIndex });
  };

  props: {
    autoStart?: boolean,
    steps: string[],
    children: any,
  };

  render() {
    const { children, steps } = this.props;
    const { tutorialIndex } = this.state;
    const { onTutorialAdvance, onTutorialDismiss, onTutorialJump } = this;

    const context = {
      tutorialIndex,
      onTutorialAdvance,
      onTutorialDismiss,
      onTutorialJump,
      tutorialSteps: steps,
    };

    return (
      <TutorialContext.Provider value={context}>
        {children}
      </TutorialContext.Provider>
    );
  }
}
