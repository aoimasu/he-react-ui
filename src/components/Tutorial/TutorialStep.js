// @flow

import classnames from 'classnames';
import * as React from 'react';
import Button from '../Form/Button';
import Icon from '../Icon';
import CarouselIndicator from '../Layout/CarouselIndicator';
import styles from './Tutorial.scss';
import withTutorial from './withTutorial';

function everyFrame(handler) {
  let stopped = false;

  function repeat() {
    if (!stopped) {
      handler();
      requestAnimationFrame(repeat);
    }
  }

  repeat();

  return function stop() {
    stopped = true;
  };
}

export default withTutorial(
  class TutorialStep extends React.Component<*, *> {
    state = {
      ownDomElement: null,
      top: 100,
      left: 76,
      reversed: false,
      isAttached: false,
    };

    componentWillMount() {
      this.stopRepeating = everyFrame(this.pollAttachedElement);
    }

    componentWillUnmount() {
      if (this.stopRepeating) this.stopRepeating();
    }

    receiveOwnDomElement = ownDomElement => {
      if (ownDomElement !== this.state.ownDomElement) {
        this.setState({ ownDomElement });
      }
    };

    pollAttachedElement = () => {
      const { attachTo } = this.props;

      const attachedDomElement = attachTo
        ? document.getElementById(attachTo)
        : null;

      if (attachedDomElement) {
        const cords = attachedDomElement.getBoundingClientRect();
        const top = cords.top + cords.height / 2;
        const left = cords.right;
        const reversed = cords.top > window.innerHeight / 2;

        if (
          top !== this.state.top ||
          left !== this.state.left ||
          reversed !== this.state.reversed
        ) {
          this.setState({ top, left, reversed, isAttached: true });
        }
      }
    };

    stopRepeating = null;

    render() {
      const {
        onTutorialAdvance,
        onTutorialDismiss,
        tutorialIndex,
        tutorialSteps,
        id,
        showCarousel,
        children,
        className,
        centered,
      } = this.props;

      const currentStep = tutorialSteps[tutorialIndex];

      if (currentStep !== id) {
        return null;
      }

      const { ownDomElement } = this.state;
      const { top, left, reversed, isAttached } = this.state;

      const classes = classnames(styles.outer, styles.showing);

      const popupClasses = classnames(styles.popup, className, {
        [styles.popupCentered]: centered,
      });
      let wrapperStyle = {
        position: 'absolute',
        width: '100%',
        height: '100%',
        // transition: 'all 0.4s',
      };
      let arrowStyle = {};
      let newTop = top - 75;

      if (ownDomElement && reversed) {
        newTop -= ownDomElement.getBoundingClientRect().height - 150;
        arrowStyle = {
          top: ownDomElement.getBoundingClientRect().height - 75,
        };
      }
      wrapperStyle = {
        ...wrapperStyle,
        top: newTop,
        left,
        right: 'auto',
      };
      const rightOverlay = {
        left,
      };
      const leftOverlay = {
        width: left,
      };

      return (
        <div className={classes}>
          <div
            className={styles.transparentOverlay}
            style={leftOverlay}
            onClick={onTutorialDismiss}
          />
          <div
            className={styles.transparentOverlay}
            style={rightOverlay}
            onClick={onTutorialDismiss}
          />
          <div style={wrapperStyle}>
            <div className={popupClasses}>
              <Icon
                className={styles.close}
                name="Cross"
                onClick={onTutorialDismiss}
              />
              <div className={styles.tutorialWrapper}>
                {children}

                {showCarousel ? (
                  <CarouselIndicator
                    className={styles.tutorialIndicator}
                    length={tutorialSteps.length}
                    current={tutorialIndex}
                    nextStep={onTutorialAdvance}
                  />
                ) : (
                  <div className={styles.tutorialIntroFooter}>
                    <div className={styles.footerCell} />
                    <div className={styles.footerCell}>
                      <Button
                        className={styles.rightElement}
                        link
                        onClick={onTutorialDismiss}
                      >
                        Got it!
                      </Button>
                    </div>
                  </div>
                )}
              </div>
            </div>
            {isAttached && (
              <Icon
                className={classnames(styles.arrow, {
                  [styles.reversed]: reversed,
                })}
                style={arrowStyle}
                name="CurvedArrow"
              />
            )}
          </div>
        </div>
      );
    }
  },
);
