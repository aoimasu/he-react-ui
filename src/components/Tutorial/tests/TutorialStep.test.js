// @flow
import { mount } from 'enzyme';
import React from 'react';
import TutorialOwner from '../TutorialOwner';
import TutorialStep from '../TutorialStep';

jest.mock('../getCoordsForElementId.js');

function makeHarness(attachTo = null) {
  return mount(
    <TutorialOwner autoStart steps={['A', 'Z']}>
      <TutorialStep id="A" showCarousel attachTo={attachTo}>
        <span id="inStepA" />
      </TutorialStep>

      <TutorialStep id="Z" showCarousel>
        <span id="inStepZ" />
      </TutorialStep>
    </TutorialOwner>,
  );
}

describe('TutorialStep', () => {
  it('Should allow advancing by clicking the next button', () => {
    const harness = makeHarness();

    const stepShown = id => harness.find(`#inStep${id}`).exists();
    expect(stepShown('A')).toBeTruthy();
    expect(stepShown('Z')).toBeFalsy();

    harness.find('.nextBtn').simulate('click');

    expect(stepShown('A')).toBeFalsy();
    expect(stepShown('Z')).toBeTruthy();

    expect(harness.find('.nextBtn').exists()).toBeFalsy();

    harness.unmount();
  });

  it('Should allow dismissing', () => {
    const harness = makeHarness();

    const stepShown = id => harness.find(`#inStep${id}`).exists();

    expect(stepShown('A')).toBeTruthy();
    expect(stepShown('Z')).toBeFalsy();

    harness.find('svg.close').simulate('click');

    expect(stepShown('A')).toBeFalsy();
    expect(stepShown('Z')).toBeFalsy();

    harness.unmount();
  });

  it('should follow its target element', () => {
    const harness = makeHarness('100-200-300');

    expect(
      harness
        .find('#inStepA')
        .closest('.outer')
        .prop('style'),
    ).toMatchSnapshot();

    harness.unmount();
  });
});
