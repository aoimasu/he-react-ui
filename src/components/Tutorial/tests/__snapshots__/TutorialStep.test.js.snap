// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TutorialStep should follow its target element 1`] = `
Object {
  "height": "100%",
  "left": 200,
  "position": "absolute",
  "right": "auto",
  "top": 175,
  "width": "100%",
}
`;
